clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -430,6 +433,33 @@ caseCon ctx e@(Case subj ty alts) = do
 
 caseCon _ e = return e
 
+
+-- | Binds variables on the PureHeap over the result of the rewrite
+--
+-- To prevent unnecessary rewrites only do this when rewrite changed something.
+bindPureHeap
+  :: TransformContext
+  -> TyConMap
+  -> PureHeap
+  -> (TransformContext -> RewriteMonad extra Term)
+  -> RewriteMonad extra Term
+bindPureHeap (TransformContext is0 ctxs) tcm heap rw = do
+  (e, Monoid.getAny -> hasChanged) <- listen $ rw ctx'
+  if hasChanged && not (null bndrs)
+    then return $ Letrec bndrs e
+    else return e
+  where
+    bndrs = map toLetBinding $ toListUniqMap heap
+    heapIds = map fst bndrs
+    is1 = extendInScopeSetList is0 heapIds
+    ctx' = TransformContext is1 (LetBody heapIds : ctxs)
+
+    toLetBinding :: (Unique,Term) -> LetBinding
+    toLetBinding (uniq,term) = (nm, term)
+      where
+        ty = termType tcm term
+        nm = mkId ty (mkUnsafeInternalName "x" uniq)
```

Comment:
Add a comment saying why it's safe to use `mkUnsafeInternalName` here; that way, if ever there's a bug caused by this line, we have an idea which assumptions were wrong.

Reply:
Done
And I changed it to `mkUnsafeSystemName`, because the original name was generated with `mkUniqSystemId`